test/rate_limiter_test.cc

```diff
@@ -68,6 +68,58 @@ TEST_F(RateLimiterTest, BurstingRateLimiterTest) {
   EXPECT_FALSE(rate_limiter->tryAcquireOne());
 }
 
+TEST_F(RateLimiterTest, ScheduledStartingRateLimiterTest) {
+  Envoy::Event::SimulatedTimeSystem time_system;
+  const auto schedule_delay = 10ms;
+  // We test regular flow, but also the flow where the first aquisition attempt comes after the
+  // scheduled delay. This should be business as usual from a functional perspective, but internally
+  // this rate limiter specializes on this case to log a warning message, and we want to cover that.
+  for (const bool starting_late : std::vector<bool>{false, true}) {
+    const Envoy::MonotonicTime scheduled_starting_time =
+        time_system.monotonicTime() + schedule_delay;
+    std::unique_ptr<MockRateLimiter> mock_rate_limiter = std::make_unique<MockRateLimiter>();
+    MockRateLimiter& unsafe_mock_rate_limiter = *mock_rate_limiter;
+    InSequence s;
+
+    EXPECT_CALL(unsafe_mock_rate_limiter, timeSource)
+        .Times(AtLeast(1))
+        .WillRepeatedly(ReturnRef(time_system));
+    RateLimiterPtr rate_limiter = std::make_unique<ScheduledStartingRateLimiter>(
+        std::move(mock_rate_limiter), scheduled_starting_time);
+    EXPECT_CALL(unsafe_mock_rate_limiter, tryAcquireOne)
+        .Times(AtLeast(1))
+        .WillRepeatedly(Return(true));
+
+    if (starting_late) {
+      time_system.sleep(schedule_delay);
+    }
+
+    // We should expect zero releases until it is time to start.
+    while (time_system.monotonicTime() < scheduled_starting_time) {
```

Comment:
If this test case doesn't succeed, it will run infinitely rather than fail.  Can you add in a tripping mechanism?

Reply:
The simulated time system here is guaranteed to move forward in this loop with `1ms` steps we add via the call to `time_system.sleep(1ms)` below.. afaict this is guaranteed to terminate?

Reply:
Nevermind here. I was overcomplicated the situation in my head. Sorry about that!